In [13]:
import pandas as pd 
import numpy as np 
#import gc
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns 
import timeit

In [15]:
#timeit setup
setup = '''
import random

random.seed('slartibartfast')
s = [random.random() for i in range(1000)]
timesort = list.sort
'''

print(min(timeit.Timer('a = s[:]; timesort(a)', 
                       setup = setup).repeat(7,1000)))

0.14642716399976052


# Read in the merged data 2.0

In [2]:
col_names = ['idx1', 'idx2', 'Id', 'EventTimeStamp', 'EquipmentID', 'DriverID', 'Latitude',
       'Longitude', 'LocationTimeStamp', 'Speed', 'Odometer', 'IgnitionStatus',
       'EFReportReason', 'TankLevelPercent', 'TankLevelGallons',
       'ExactFuelEventId', 'truck', 'tankcap', 'refuel_datetime',
       'refuel_tr_sum', 'refuel_tr_id']
date_cols = ['EventTimeStamp', 'LocationTimeStamp', 'refuel_datetime' ]

In [3]:
dtype_dict = {
    'Id' : 'uint64', 
    'EquipmentID' : 'category', 
    'DriverID' : 'category', 
    'Latitude' : 'float64', 
    'Longitude' : 'float64', 
    'Speed' : 'uint16', 
    'Odometer' : 'float64', 
    'IgnitionStatus' : 'category', 
    'EFReportReason' : 'category', 
    'TankLevelPercent' : 'float32', 
    'TankLevelGallons' : 'float32', 
    'ExactFuelEventId' : 'uint64', 
    'truck' : 'category', 
    'tankcap' : 'uint16', 
    'refuel_tr_sum' : 'float32', 
    'refuel_tr_id' : 'category' }

In [4]:
## Read the csv after creating the dtype_dict, col_names, and date_cols objects
events_levels_df = pd.read_csv('/Users/paulomartinez/Downloads/events_levels_df.gz', skiprows=1, names=col_names,
                               dtype=dtype_dict, usecols=np.arange(2,19), 
                              parse_dates = date_cols, infer_datetime_format = True)
#11:20 pm - 11:21 pm
#11:30 pm - 11:31 pm

In [5]:
events_levels_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7908406 entries, 0 to 7908405
Data columns (total 17 columns):
Id                   uint64
EventTimeStamp       datetime64[ns]
EquipmentID          category
DriverID             category
Latitude             float64
Longitude            float64
LocationTimeStamp    datetime64[ns]
Speed                uint16
Odometer             float64
IgnitionStatus       category
EFReportReason       category
TankLevelPercent     float32
TankLevelGallons     float32
ExactFuelEventId     uint64
truck                category
tankcap              uint16
refuel_datetime      datetime64[ns]
dtypes: category(5), datetime64[ns](3), float32(2), float64(3), uint16(2), uint64(2)
memory usage: 633.6 MB


In [6]:
events_levels_df.head()

Id          EventTimeStamp EquipmentID DriverID  Latitude  Longitude  \
0  12649676 2017-01-01 00:19:45.357        1559      NaN  36.06712  -86.43476   
1  12649677 2017-01-01 00:34:45.370        1559      NaN  36.06712  -86.43476   
2  12649678 2017-01-01 00:49:45.417        1559      NaN  36.06712  -86.43476   
3  12649679 2017-01-01 01:04:45.463        1559      NaN  36.06712  -86.43476   
4  12649680 2017-01-01 01:19:45.480        1559      NaN  36.06712  -86.43476   

    LocationTimeStamp  Speed  Odometer IgnitionStatus EFReportReason  \
0 2017-01-01 00:01:05      0  505629.3              1              2   
1 2017-01-01 00:01:05      0  505629.3              1              2   
2 2017-01-01 00:01:05      0  505629.3              1              2   
3 2017-01-01 00:01:05      0  505629.3              1              2   
4 2017-01-01 00:01:05      0  505629.3              1              2   

   TankLevelPercent  TankLevelGallons  ExactFuelEventId truck  tankcap  \
0              96.0             240.0          12649676  1559      250   
1              96.0             240.0          12649677  1559      250   
2              96.0             240.0          12649678  1559      250   
3              96.0             240.0          12649679  1559      250   
4              96.0             240.0          12649680  1559      250   

  refuel_datetime  
0             NaT  
1             NaT  
2             NaT  
3             NaT  
4             NaT

# Drop the noisy rows

Ok, let's try an alternative.
1. drop superfluous columns
2. sort by truck and datetime
3. run flag function
4. define slice
    - remove noisy df

## --- lets first make sure the df isn't corrupted somehow

In [7]:
list(events_levels_df.columns.values)

['Id',
 'EventTimeStamp',
 'EquipmentID',
 'DriverID',
 'Latitude',
 'Longitude',
 'LocationTimeStamp',
 'Speed',
 'Odometer',
 'IgnitionStatus',
 'EFReportReason',
 'TankLevelPercent',
 'TankLevelGallons',
 'ExactFuelEventId',
 'truck',
 'tankcap',
 'refuel_datetime']

In [8]:
for col in list(events_levels_df.columns.values):
    print('len of set of col:', str(col), 
          len(set(events_levels_df[str(col)]))
         )

len of set of col: Id 7908406
len of set of col: EventTimeStamp 7527560
len of set of col: EquipmentID 629
len of set of col: DriverID 925
len of set of col: Latitude 1032388
len of set of col: Longitude 1286842
len of set of col: LocationTimeStamp 6551558
len of set of col: Speed 88
len of set of col: Odometer 3002046
len of set of col: IgnitionStatus 2
len of set of col: EFReportReason 3
len of set of col: TankLevelPercent 1000
len of set of col: TankLevelGallons 4333
len of set of col: ExactFuelEventId 7908406
len of set of col: truck 629
len of set of col: tankcap 5
len of set of col: refuel_datetime 89976


In [9]:
#1. drop superfluous columns
events_levels_df.drop('EquipmentID', inplace=True, axis='columns')
events_levels_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7908406 entries, 0 to 7908405
Data columns (total 16 columns):
Id                   uint64
EventTimeStamp       datetime64[ns]
DriverID             category
Latitude             float64
Longitude            float64
LocationTimeStamp    datetime64[ns]
Speed                uint16
Odometer             float64
IgnitionStatus       category
EFReportReason       category
TankLevelPercent     float32
TankLevelGallons     float32
ExactFuelEventId     uint64
truck                category
tankcap              uint16
refuel_datetime      datetime64[ns]
dtypes: category(4), datetime64[ns](3), float32(2), float64(3), uint16(2), uint64(2)
memory usage: 618.5 MB


In [10]:
events_levels_df.head()

Id          EventTimeStamp DriverID  Latitude  Longitude  \
0  12649676 2017-01-01 00:19:45.357      NaN  36.06712  -86.43476   
1  12649677 2017-01-01 00:34:45.370      NaN  36.06712  -86.43476   
2  12649678 2017-01-01 00:49:45.417      NaN  36.06712  -86.43476   
3  12649679 2017-01-01 01:04:45.463      NaN  36.06712  -86.43476   
4  12649680 2017-01-01 01:19:45.480      NaN  36.06712  -86.43476   

    LocationTimeStamp  Speed  Odometer IgnitionStatus EFReportReason  \
0 2017-01-01 00:01:05      0  505629.3              1              2   
1 2017-01-01 00:01:05      0  505629.3              1              2   
2 2017-01-01 00:01:05      0  505629.3              1              2   
3 2017-01-01 00:01:05      0  505629.3              1              2   
4 2017-01-01 00:01:05      0  505629.3              1              2   

   TankLevelPercent  TankLevelGallons  ExactFuelEventId truck  tankcap  \
0              96.0             240.0          12649676  1559      250   
1              96.0             240.0          12649677  1559      250   
2              96.0             240.0          12649678  1559      250   
3              96.0             240.0          12649679  1559      250   
4              96.0             240.0          12649680  1559      250   

  refuel_datetime  
0             NaT  
1             NaT  
2             NaT  
3             NaT  
4             NaT

In [17]:
#sort by truck and date
events_levels_df.sort_values(['truck', 'EventTimeStamp'],
                                     inplace = True
                                         )
events_levels_df
#started sort at 12:10 am - 12:10 am

Id          EventTimeStamp DriverID  Latitude  Longitude  \
2443    12804900 2017-01-09 08:03:30.687      NaN  36.06612  -86.43410   
2445    12804901 2017-01-09 08:05:30.670      NaN  36.06612  -86.43410   
2451    12804902 2017-01-09 08:17:10.670      NaN  36.06612  -86.43410   
2514    12804904 2017-01-09 12:24:31.170      NaN  36.06612  -86.43410   
2515    12804903 2017-01-09 12:25:41.203      NaN  36.06612  -86.43410   
2530    12804905 2017-01-09 14:04:41.767      NaN  36.06612  -86.43410   
2531    12804906 2017-01-09 14:06:31.780      NaN  36.06612  -86.43410   
2535    12804907 2017-01-09 14:26:11.797      NaN  36.06612  -86.43410   
2536    12804908 2017-01-09 14:28:21.813      NaN  36.06612  -86.43410   
2543    12804899 2017-01-09 14:36:01.877      NaN  36.06612  -86.43410   
2545    12804909 2017-01-09 14:38:51.890      NaN  36.06612  -86.43410   
2547    12804910 2017-01-09 14:40:51.890      NaN  36.06612  -86.43410   
2555    12804911 2017-01-09 15:00:41.920      NaN  36.06612  -86.43410   
2564    12804912 2017-01-09 15:15:51.920      NaN  36.06612  -86.43410   
2571    12804913 2017-01-09 15:24:31.920      NaN  36.06612  -86.43410   
406354  12650361 2017-01-01 08:12:27.390    BAIJO  39.69715  -86.18514   
406355  12650362 2017-01-01 08:14:27.390    BAIJO  39.69715  -86.18514   
406356  12650363 2017-01-01 08:29:37.403    BAIJO  39.69715  -86.18514   
406357  12650364 2017-01-01 08:30:47.403    BAIJO  39.69715  -86.18514   
406358  12651601 2017-01-01 12:05:57.983    BAIJO  39.69725  -86.18459   
406359  12651602 2017-01-01 12:07:47.967    BAIJO  39.69759  -86.18493   
406360  12651603 2017-01-01 12:08:18.090    BAIJO  39.69759  -86.18493   
406361  12651635 2017-01-01 12:13:27.997    BAIJO  39.69734  -86.18494   
406362  12651636 2017-01-01 12:15:37.997    BAIJO  39.69806  -86.18634   
406363  12651716 2017-01-01 12:30:38.060    BAIJO  39.60912  -86.07220   
406364  12652008 2017-01-01 13:16:28.233    BAIJO  38.95619  -85.85215   
406365  12652047 2017-01-01 13:18:28.233    BAIJO  38.95356  -85.85320   
406366  12652048 2017-01-01 13:20:28.263    BAIJO  38.95070  -85.85023   
406367  12652072 2017-01-01 13:22:28.263    BAIJO  38.95036  -85.84935   
406368  12652109 2017-01-01 13:26:28.263    BAIJO  38.95098  -85.84871   
...          ...                     ...      ...       ...        ...   
673701  21118922 2018-03-21 22:29:15.637     BEVL  35.46718  -86.45137   
673702  21118936 2018-03-21 22:29:15.637     BEVL  35.46718  -86.45137   
673703  21119031 2018-03-21 22:44:15.747     BEVL  35.31484  -86.35906   
673704  21119063 2018-03-21 22:44:15.747     BEVL  35.31484  -86.35906   
673705  21119100 2018-03-21 22:54:05.873     BEVL  35.26450  -86.39860   
673706  21119154 2018-03-21 22:54:05.873     BEVL  35.26450  -86.39860   
673707  21119224 2018-03-21 23:04:45.920     BEVL  35.26409  -86.39845   
673708  21119340 2018-03-21 23:19:45.950     BEVL  35.26410  -86.39848   
673709  21119397 2018-03-21 23:34:46.030     BEVL  35.27240  -86.39742   
673710  21119416 2018-03-21 23:38:06.013     BEVL  35.26360  -86.39919   
673711  21119458 2018-03-21 23:46:26.043     BEVL  35.26241  -86.39509   
673712  21119532 2018-03-22 00:01:26.137     BEVL  35.20152  -86.49736   
673713  21119618 2018-03-22 00:16:36.263     BEVL  35.14628  -86.57983   
673714  21119689 2018-03-22 00:18:46.280     BEVL  35.14272  -86.58209   
673715  21119706 2018-03-22 00:18:46.280     BEVL  35.14272  -86.58209   
673716  21119690 2018-03-22 00:29:56.327     BEVL  35.15319  -86.59292   
673717  21119707 2018-03-22 00:29:56.327     BEVL  35.15319  -86.59292   
673718  21119721 2018-03-22 00:35:06.373     BEVL  35.15119  -86.59468   
673719  21119823 2018-03-22 00:50:06.437     BEVL  35.20706  -86.48150   
673720  21119905 2018-03-22 01:05:06.513     BEVL  35.27241  -86.39709   
673721  21119925 2018-03-22 01:07:26.543     BEVL  35.27230  -86.39701   
673722  21120012 2018-03-22 01:23:46.717     BEVL  35.27230  -86.39748   
673723  21120101 2

In [18]:
#reset the index
events_levels_df.reset_indext_index(inplace = True, drop = True)
events_levels_df

Id          EventTimeStamp DriverID  Latitude  Longitude  \
0        12804900 2017-01-09 08:03:30.687      NaN  36.06612  -86.43410   
1        12804901 2017-01-09 08:05:30.670      NaN  36.06612  -86.43410   
2        12804902 2017-01-09 08:17:10.670      NaN  36.06612  -86.43410   
3        12804904 2017-01-09 12:24:31.170      NaN  36.06612  -86.43410   
4        12804903 2017-01-09 12:25:41.203      NaN  36.06612  -86.43410   
5        12804905 2017-01-09 14:04:41.767      NaN  36.06612  -86.43410   
6        12804906 2017-01-09 14:06:31.780      NaN  36.06612  -86.43410   
7        12804907 2017-01-09 14:26:11.797      NaN  36.06612  -86.43410   
8        12804908 2017-01-09 14:28:21.813      NaN  36.06612  -86.43410   
9        12804899 2017-01-09 14:36:01.877      NaN  36.06612  -86.43410   
10       12804909 2017-01-09 14:38:51.890      NaN  36.06612  -86.43410   
11       12804910 2017-01-09 14:40:51.890      NaN  36.06612  -86.43410   
12       12804911 2017-01-09 15:00:41.920      NaN  36.06612  -86.43410   
13       12804912 2017-01-09 15:15:51.920      NaN  36.06612  -86.43410   
14       12804913 2017-01-09 15:24:31.920      NaN  36.06612  -86.43410   
15       12650361 2017-01-01 08:12:27.390    BAIJO  39.69715  -86.18514   
16       12650362 2017-01-01 08:14:27.390    BAIJO  39.69715  -86.18514   
17       12650363 2017-01-01 08:29:37.403    BAIJO  39.69715  -86.18514   
18       12650364 2017-01-01 08:30:47.403    BAIJO  39.69715  -86.18514   
19       12651601 2017-01-01 12:05:57.983    BAIJO  39.69725  -86.18459   
20       12651602 2017-01-01 12:07:47.967    BAIJO  39.69759  -86.18493   
21       12651603 2017-01-01 12:08:18.090    BAIJO  39.69759  -86.18493   
22       12651635 2017-01-01 12:13:27.997    BAIJO  39.69734  -86.18494   
23       12651636 2017-01-01 12:15:37.997    BAIJO  39.69806  -86.18634   
24       12651716 2017-01-01 12:30:38.060    BAIJO  39.60912  -86.07220   
25       12652008 2017-01-01 13:16:28.233    BAIJO  38.95619  -85.85215   
26       12652047 2017-01-01 13:18:28.233    BAIJO  38.95356  -85.85320   
27       12652048 2017-01-01 13:20:28.263    BAIJO  38.95070  -85.85023   
28       12652072 2017-01-01 13:22:28.263    BAIJO  38.95036  -85.84935   
29       12652109 2017-01-01 13:26:28.263    BAIJO  38.95098  -85.84871   
...           ...                     ...      ...       ...        ...   
7908376  21118922 2018-03-21 22:29:15.637     BEVL  35.46718  -86.45137   
7908377  21118936 2018-03-21 22:29:15.637     BEVL  35.46718  -86.45137   
7908378  21119031 2018-03-21 22:44:15.747     BEVL  35.31484  -86.35906   
7908379  21119063 2018-03-21 22:44:15.747     BEVL  35.31484  -86.35906   
7908380  21119100 2018-03-21 22:54:05.873     BEVL  35.26450  -86.39860   
7908381  21119154 2018-03-21 22:54:05.873     BEVL  35.26450  -86.39860   
7908382  21119224 2018-03-21 23:04:45.920     BEVL  35.26409  -86.39845   
7908383  21119340 2018-03-21 23:19:45.950     BEVL  35.26410  -86.39848   
7908384  21119397 2018-03-21 23:34:46.030     BEVL  35.27240  -86.39742   
7908385  21119416 2018-03-21 23:38:06.013     BEVL  35.26360  -86.39919   
7908386  21119458 2018-03-21 23:46:26.043     BEVL  35.26241  -86.39509   
7908387  21119532 2018-03-22 00:01:26.137     BEVL  35.20152  -86.49736   
7908388  21119618 2018-03-22 00:16:36.263     BEVL  35.14628  -86.57983   
7908389  21119689 2018-03-22 00:18:46.280     BEVL  35.14272  -86.58209   
7908390  21119706 2018-03-22 00:18:46.280     BEVL  35.14272  -86.58209   
7908391  21119690 2018-03-22 00:29:56.327     BEVL  35.15319  -86.59292   
7908392  21119707 2018-03-22 00:29:56.327     BEVL  35.15319  -86.59292   
7908393  21119721 2018-03-22 00:35:06.373     BEVL  35.15119  -86.59468   
7908394  21119823 2018-03-22 00:50:06.437     BEVL  35.20706  -86.48150   
7908395  21119905 2018-03-22 01:05:06.513     BEVL  35.27241  -86.39709   
7908396  21119925 2018-03-22 01:07:26.543     BEVL  35.27230  -86.39701   
7908397  21120012 2018-03-22 01:23:46.71

In [22]:
def flag_anomal_odom_rows_2(df):#!!!! requires a dataframe with a reset index!!!!
    '''takes a dataframe with an Odometer column and returns a boolean index of 
    identifying which rows are noisy Odometer readings and should be dropped.
    '''
    #initialize a series, without attaching it to the df in the hopes of expediency
    OdomDiff = df.Odometer.diff()
    #initialize list which will serve as a boolean index
    flags = []
    flag = 1 #this will be flipped like a switch 1 means do not drop, 0 means drop
    #loop through each difference measurement
    for i in range(len(OdomDiff) - 1):
    # the loop will flag all but the last row because otherwise I'd have to find a way to loop around the "corner" of the index
        #will start by flaggin with 1 (do not drop) until an anomaly flips it.
        flags.append(bool(flag))
    #   # for each truck, if odometer change is anomalous (i.e. spikes up or down) flip the flag switch
        if (df.truck[i + 1] == df.truck[i]) & (abs(OdomDiff[i + 1]) > 200): 
            flag = (flag + 1)%2
    #       #the flag will remain on the delete (i.e. 0) setting until the loop detects the spike that corrects the odometer
    #   # if we change trucks before the anomaly is corrected reset the flag to do not drop (i.e. 1)   
        if df.truck[i + 1] != df.truck[i]:
            flag = 1
    #now that the loop is done, we append the last flag (which is in the state the loop left it. So, if a loop ends before the anomaly is corrected, the flag will still be set at 0/False)
    flags.append(bool(flag))
    return flags

In [ ]:
t0 = datetime.now()
print('started at ', datetime.now())
flags = flag_anomal_odom_rows_2(events_levels_df)
tf = datetime.now()
print('finished at ', tf)
print('whole process took ', tf-t0)
    #eliminated the print statement, and some empty lines
    #ran second attempt at 5/4/18, 00:41 am CST

    #the print statement I had in the function was taking too long.
    # I'm getting about 3 rows/second
        # => 2,650,000 seconds = 44, 167 minutes = 736 hours = 31 days... 😒 

started at  2018-05-04 03:43:05.333591


In [ ]:
#alright, assuming this works I'll want to redefine the df by dropping the noisy rows
print('started at ', datetime.now())
events_levels_df = events_levels_df[flags]
tf = datetime.now()
print('finished at ', tf)
print('whole process took ', tf-t0)
events_levels_df.info()

# Assuming row cleaning was successful 🤞

#### Now I need to drop all rows with timestamps inside of the time range of consecutive refuelings
- to do this, I'll need to identify each trucks fueling purchases

It looks like Jarrod’s merge produced a litany of duplicate refuel_datetime values. ⛽️ ⛽️ ...⛽️ 

- ✅ This could make sense since every fuel level measurement that happened between refuel_datetime times might have received the preceding refuel_datetime time-stamp.

Assuming this is the case, I should be able to run a modified version of my flag function to flag "inter-fueling-rows". 🇲🇽 
- leaving us with a "feather-light" 89,976 rows


##### attempt to flag inter-fueling-rows

    🦂 
    (is it possible that I would have dropped a noisy row that also happened to identify the next fueling event? Perhaps, (although the odds of having this problem are only 411/7.9Million). Regardless, we should be fine, since the following row will have the appropriate timestamp and odometer reading.) 
    So, we're good to go 
    😁 🚀 

In [ ]:
#reset the index now that we have dropped the noisy odometer rows
events_levels_df.reset_indext_index(inplace = True, drop = True)
events_levels_df